File: Medium/2026-04-07-2069-Walking-Robot-Simulation-II/solution.py
class Robot:
    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        self.x = 0
        self.y = 0
        self.directions = ["East", "North", "West", "South"]
        self.direction_index = 0  # Initially facing East

    def step(self, num: int) -> None:
        for _ in range(num):
            if self.direction_index == 0:  # East
                if self.x + 1 < self.width:
                    self.x += 1
                else:
                    self.direction_index = 1  # Turn North
            elif self.direction_index == 1:  # North
                if self.y + 1 < self.height:
                    self.y += 1
                else:
                    self.direction_index = 2  # Turn West
            elif self.direction_index == 2:  # West
                if self.x - 1 >= 0:
                    self.x -= 1
                else:
                    self.direction_index = 3  # Turn South
            elif self.direction_index == 3:  # South
                if self.y - 1 >= 0:
                    self.y -= 1
                else:
                    self.direction_index = 0  # Turn East

    def getPos(self) -> list[int]:
        return [self.x, self.y]

    def getDir(self) -> str:
        return self.directions[self.direction_index]

File: Medium/2026-04-07-2069-Walking-Robot-Simulation-II/solution.java
class Robot {
    private int x, y, width, height, direction;
    private final int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}; // North, East, South, West

    public Robot(int width, int height) {
        this.width = width;
        this.height = height;
        this.x = 0;
        this.y = 0;
        this.direction = 1; // Initially facing East
    }

    public void step(int num) {
        for (int i = 0; i < num; i++) {
            int newX = x + directions[direction][0];
            int newY = y + directions[direction][1];
            if (newX < 0 || newX >= width || newY < 0 || newY >= height) {
                direction = (direction + 3) % 4; // Turn 90 degrees counterclockwise
            } else {
                x = newX;
                y = newY;
            }
        }
    }

    public int[] getPos() {
        return new int[]{x, y};
    }

    public String getDir() {
        switch (direction) {
            case 0: return "North";
            case 1: return "East";
            case 2: return "South";
            case 3: return "West";
            default: return "";
        }
    }
}